lib/hanami/model/sql/types/schema/coercions.rb

```diff
@@ -192,6 +192,12 @@ def self.hash(arg)
                 raise ArgumentError.new("invalid value for Hash(): #{arg.inspect}")
               end
             end
+
+            def self.json(arg)
```

Comment:
@jodosha It's almost certain that `Types::Hash.constructor(Coercions.method(:json))` will receive an array at some point and that's why I decided to write this `json` coercion in the first place.

The coercion tries to convert input to `Hash` and only when it fails it tries to convert to `Array`. And I lazily used exception for that to avoid copying coercion logic, but you are right that it's not a good way to go.